src/Symfony/Component/Security/Http/Authenticator/MagicLinkAuthenticator.php

```diff
@@ -0,0 +1,87 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Authenticator;
+
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\InvalidMagicLinkAuthenticationException;
+use Symfony\Component\Security\Http\Authentication\AuthenticationFailureHandlerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+use Symfony\Component\Security\Http\HttpUtils;
+use Symfony\Component\Security\Http\MagicLink\Exception\InvalidMagicLinkExceptionInterface;
+use Symfony\Component\Security\Http\MagicLink\MagicLoginLinkerInterface;
+
+/**
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ */
+class MagicLinkAuthenticator extends AbstractAuthenticator implements InteractiveAuthenticatorInterface
+{
+    private $magicLoginLinkHelper;
+    private $httpUtils;
+    private $successHandler;
+    private $failureHandler;
+    private $options;
+
+    public function __construct(MagicLoginLinkerInterface $magicLoginLinkHelper, HttpUtils $httpUtils, AuthenticationSuccessHandlerInterface $successHandler, AuthenticationFailureHandlerInterface $failureHandler, array $options)
+    {
+        $this->magicLoginLinkHelper = $magicLoginLinkHelper;
+        $this->httpUtils = $httpUtils;
+        $this->successHandler = $successHandler;
+        $this->failureHandler = $failureHandler;
+        $this->options = $options;
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return $this->httpUtils->checkRequestPath($request, $this->options['check_route']);
+    }
+
+    public function authenticate(Request $request): PassportInterface
+    {
+        $uri = $request->getUri();
+
+        return new SelfValidatingPassport(
+            new UserBadge($uri, function ($uri) {
+                try {
+                    $user = $this->magicLoginLinkHelper
+                        ->consumeMagicLink($uri);
+                } catch (InvalidMagicLinkExceptionInterface $e) {
+                    throw new InvalidMagicLinkAuthenticationException('Magic link could not be validated.', 0, $e);
+                }
+
+                return $user;
+            }),
+            []
+        );
+    }
+
+    public function onAuthenticationSuccess(Request $request, TokenInterface $token, string $firewallName): ?Response
+    {
+        return new RedirectResponse($this->httpUtils->generateUri($request, 'app_homepage'));
```

Comment:
Shouldn't this use the `successHandler` instead of this "random" default route?

Reply:
Ha! Thanks for catching that - was a total oversight :)